src/callbacks_step/analysis_dg2d.jl

```diff
@@ -54,13 +54,33 @@ end
 function calc_error_norms(func, u, t, analyzer,
                           mesh::TreeMesh{2}, equations, initial_condition,
                           dg::DGSEM, cache, cache_analysis)
+  # Calculate error norms for each element
+  l2_errors, linf_errors = calc_error_norms_per_element(func, u, t, analyzer,
+                                                        mesh, equations, initial_condition,
+                                                        dg, cache, cache_analysis)
+
+  # Aggregate element error norms
+  l2_error = sum(l2_errors)
+  linf_error = reduce((x, y) -> max.(x, y), linf_errors)
+
+  # For L2 error, divide by total volume
+  total_volume_ = total_volume(mesh)
+  l2_error = @. sqrt(l2_error / total_volume_)
+
+  return l2_error, linf_error
+end
+
+function calc_error_norms_per_element(func, u, t, analyzer,
+                                      mesh::TreeMesh{2}, equations, initial_condition,
+                                      dg::DGSEM, cache, cache_analysis)
   @unpack vandermonde, weights = analyzer
   @unpack node_coordinates = cache.elements
   @unpack u_local, u_tmp1, x_local, x_tmp1 = cache_analysis
 
   # Set up data structures
-  l2_error   = zero(func(get_node_vars(u, equations, dg, 1, 1, 1), equations))
-  linf_error = copy(l2_error)
+  T = typeof(zero(func(get_node_vars(u, equations, dg, 1, 1, 1), equations)))
+  l2_errors = zeros(T, nelements(dg, cache))
+  linf_errors = copy(l2_errors)
```

Comment:
Would it make sense to store these arrays in the `cache_analysis` to avoid allocations during execution?

Reply:
I'm not sure about this. During development, the performance wasn't noticeably worse than before. But I didn't do any benchmarks to get some exact numbers. 

Would it be easy to store these arrays in the cache? Maybe I'm missing something here, but I think the `eltype` of them depends on `func` so it's probably not straightforward to implement this.

Reply:
@sloede I have rewritten the serial error calculation in 11f57fd as discussed. It doesn't allocate anymore and the order of summation is still the same as in the parallel case.